src/main/java/org/folio/service/UserService.java

```diff
@@ -0,0 +1,161 @@
+          if (users.isEmpty()) {
+            String message = "No user found by " + userPropertyName + " == " + samlAttributeValue;
+            throw new UserErrorException(message);
+          }
+          return users.getJsonObject(0);
+        }));
+  }
+
+  private Future<String> handleGetUser(String userPropertyName, String value, WebClient webClient, OkapiHeaders okapiHeaders) {
+    return getUserTenant(userPropertyName, value, webClient, okapiHeaders)
+      .compose(userTenantsObject -> {
+        if (userTenantsObject.getInteger(TOTAL_RECORDS) == 0) {
+          return Future.succeededFuture(okapiHeaders.getTenant());
+        } else if (userTenantsObject.getInteger(TOTAL_RECORDS) > 1) {
+          log.warn(MULTIPLE_MATCHING_USERS, userPropertyName, value);
+          return Future.succeededFuture(okapiHeaders.getTenant());
+        } else {
+          String tenantId = userTenantsObject.getJsonArray(USER_TENANTS).getJsonObject(0).getString(TENANT_ID);
+          if (StringUtils.isBlank(tenantId)) {
+            String errorMassage = String.format(USER_TENANT_DOES_NOT_HAVE_A_TENANT_ID_ERROR, userTenantsObject, value);
+            log.error("handleGetUser: {}", errorMassage);
+            return Future.failedFuture(errorMassage);
+          }
+          okapiHeaders.setTenant(tenantId);
+          return Future.succeededFuture(tenantId);
+        }
+      });
+  }
+
+  private Future<JsonObject> getUserTenant(String userPropertyName, String value, WebClient webClient, OkapiHeaders okapiHeaders) {
+    final String userTenantQuery = okapiHeaders.getUrl() + USER_TENANT_ENDPOINT;
+    HttpRequest<Buffer> request = webClient.getAbs(userTenantQuery);
+
+    switch (userPropertyName) {
+      case BARCODE -> request.addQueryParam(BARCODE, value);
+      case EXTERNAL_SYSTEM_ID -> request.addQueryParam(EXTERNAL_SYSTEM_ID, value);
+      case ID -> request.addQueryParam(USER_ID, value);
+      case USERNAME -> request.addQueryParam(USERNAME, value);
+      case PERSONAL_EMAIL -> request.addQueryParam(EMAIL, value);
+    }
+
+    return request.putHeader(XOkapiHeaders.TOKEN, okapiHeaders.getToken())
+      .putHeader(XOkapiHeaders.TENANT, okapiHeaders.getTenant())
+      .expect(ResponsePredicate.SC_OK)
+      .expect(ResponsePredicate.JSON)
+      .send()
+      .map(HttpResponse::bodyAsJsonObject)
+      .recover(e -> {
+        String message = String.format(USER_TENANT_GET_ERROR, userPropertyName, value, e.getMessage());
+        log.info("{}", message, e);
+        return Future.failedFuture(message);
+      });
+  }
+
+  /**
+   * Get the user id from the first samlAttribute of userProfile.
+   *
+   * @param samlAttribute attribute name, or null for "UserID"
+   */
+  public static String getSamlAttributeValue(String samlAttribute, UserProfile userProfile) {
+    String samlAttributeName = samlAttribute == null ? "UserID" : samlAttribute;
+    List<?> samlAttributeList = getList(userProfile.getAttribute(samlAttributeName));
```

Comment:
**Minor/Question**
Maybe it's better to inline `getList`
```
    var attribute = userProfile.getAttribute(samlAttributeName);
    if (attribute == null || attribute instanceof List<?> list && list.isEmpty()) {
      throw new UserErrorException("SAML attribute doesn't exist: " + samlAttributeName);
    }
    return attribute instanceof List<?> list ? list.get(0).toString() : attribute.toString();
```

Reply:
This is not my implementation, I just moved it from the controller.